pkg/handlers/publicapi/transportation_service_provider.go

```diff
@@ -0,0 +1,81 @@
+package publicapi
+
+import (
+	"github.com/go-openapi/runtime/middleware"
+	"github.com/go-openapi/strfmt"
+	"github.com/gofrs/uuid"
+	"go.uber.org/zap"
+
+	"github.com/transcom/mymove/pkg/auth"
+	"github.com/transcom/mymove/pkg/gen/apimessages"
+	tspop "github.com/transcom/mymove/pkg/gen/restapi/apioperations/transportation_service_provider"
+	"github.com/transcom/mymove/pkg/handlers"
+	"github.com/transcom/mymove/pkg/models"
+)
+
+func payloadForTransportationServieProviderModel(t models.TransportationServiceProvider) *apimessages.TransportationServiceProvider {
+	transporationServiceProviderPayload := &apimessages.TransportationServiceProvider{
+		ID:                       *handlers.FmtUUID(t.ID),
+		StandardCarrierAlphaCode: handlers.FmtString(t.StandardCarrierAlphaCode),
+		CreatedAt:                strfmt.DateTime(t.CreatedAt),
+		UpdatedAt:                strfmt.DateTime(t.UpdatedAt),
+		Enrolled:                 t.Enrolled,
+		Name:                     t.Name,
+		PocGeneralName:           t.PocGeneralName,
+		PocGeneralEmail:          t.PocGeneralEmail,
+		PocGeneralPhone:          t.PocGeneralPhone,
+		PocClaimsName:            t.PocClaimsName,
+		PocClaimsEmail:           t.PocClaimsEmail,
+		PocClaimsPhone:           t.PocClaimsPhone,
+	}
+	return transporationServiceProviderPayload
+}
+
+// GetTransportationServiceProviderHandler returns a TSP for a shipment
+type GetTransportationServiceProviderHandler struct {
+	handlers.HandlerContext
+}
+
+// Handle getting the tsp for a shipment
+func (h GetTransportationServiceProviderHandler) Handle(params tspop.GetTransportationServiceProviderParams) middleware.Responder {
+	var shipment *models.Shipment
+	var err error
+	session := auth.SessionFromRequestContext(params.HTTPRequest)
+	shipmentID, _ := uuid.FromString(params.ShipmentID.String())
+
+	if session.IsTspUser() {
+		// TODO (2018_08_27 cgilmer): Find a way to check Shipment belongs to TSP without 2 queries
+		tspUser, err := models.FetchTspUserByID(h.DB(), session.TspUserID)
+		if err != nil {
+			h.Logger().Error("DB Query", zap.Error(err))
+			// return tspop.NewGetTransporationServiceProviderForbidden()
+		}
+
+		shipment, err = models.FetchShipmentByTSP(h.DB(), tspUser.TransportationServiceProviderID, shipmentID)
+		if err != nil {
+			h.Logger().Error("DB Query", zap.Error(err))
+			return tspop.NewGetTransportationServiceProviderBadRequest()
+		}
+	} else if session.IsOfficeUser() {
+		shipment, err = models.FetchShipment(h.DB(), session, shipmentID)
+		if err != nil {
+			h.Logger().Error("DB Query", zap.Error(err))
+			return tspop.NewGetTransportationServiceProviderBadRequest()
+		}
+
+	} else {
+		return tspop.NewGetTransportationServiceProviderForbidden()
+	}
+
+	// Assume that the last shipmentOffer contains the current TSP
+	// This might be a bad assumption, but TSPs can't currently reject offers
+	lastItemIndex := len(shipment.ShipmentOffers) - 1
```

Comment:
I think some of this code could be moved to a function in `models` so that this handler doesn't need to deal with ShipmentOffers:

```suggestion
	tsp, err := models.AcceptedTSPForShipment(h.DB(), shipmentId)
	if err != nil {
	  return handlers.ResponseForError(h.Logger(), err)
	}
	tspPayload := payloadForTSPModel(*tsp)
	return tspop.NewGetTransportationServiceProviderOK().WithPayload(tspPayload)
```

Reply:
I kind of split the difference here. The shipment model returns the id of the current TSP, and then the handler fetches it using that id. That made the most sense to me, in the case where we want to add some auth around the tsp, but happy to change further if you think it's needed.